pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,337 @@
+	logger             *zap.Logger
+}
+
+func NewReporter(
+	apiClient apiClient,
+	gitClient gitClient,
+	appLister applicationLister,
+	envGetter environmentGetter,
+	cfg *config.PipedSpec,
+	gracePeriod time.Duration,
+	logger *zap.Logger,
+) *Reporter {
+	return &Reporter{
+		apiClient:          apiClient,
+		gitClient:          gitClient,
+		applicationLister:  appLister,
+		envGetter:          envGetter,
+		config:             cfg,
+		gracePeriod:        gracePeriod,
+		lastScannedCommits: make(map[string]string),
+		// FIXME: Find another way to specify the root dir for a file system
+		fsys:   os.DirFS("/"),
+		logger: logger.Named("app-config-reporter"),
+	}
+}
+
+func (r *Reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app-config-reporter")
+
+	// Pre-clone to cache the registered git repositories.
+	r.gitRepos = make(map[string]git.Repo, len(r.config.Repositories))
+	for _, repoCfg := range r.config.Repositories {
+		repo, err := r.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+		if err != nil {
+			r.logger.Error("failed to clone repository",
+				zap.String("repo-id", repoCfg.RepoID),
+				zap.Error(err),
+			)
+			return err
+		}
+		r.gitRepos[repoCfg.RepoID] = repo
+	}
+
+	// FIXME: Think about sync interval of app config reporter
+	ticker := time.NewTicker(r.config.SyncInterval.Duration())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.scanAppConfigs(ctx); err != nil {
+				r.logger.Error("failed to check application configurations defined in Git", zap.Error(err))
+			}
+		case <-ctx.Done():
+			r.logger.Info("app-config-reporter has been stopped")
+			return nil
+		}
+	}
+}
+
+// scanAppConfigs checks and reports two types of applications.
+// One is applications registered in Control-plane already, and another is ones that aren't registered yet.
+func (r *Reporter) scanAppConfigs(ctx context.Context) error {
+	if len(r.gitRepos) == 0 {
+		r.logger.Info("no repositories were configured for this piped")
+		return nil
+	}
+
+	// Make all repos up-to-date.
+	for repoID, repo := range r.gitRepos {
+		if err := repo.Pull(ctx, repo.GetClonedBranch()); err != nil {
+			r.logger.Error("failed to update repo to latest",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			return err
+		}
+	}
+
+	// Create a map to determine from GitPath if the application is registered.
+	apps := r.applicationLister.List()
+	registeredAppPaths := make(map[string]struct{}, len(apps))
+	for _, app := range apps {
+		id := model.BuildGitPathID(app.GitPath.Repo.Id, app.GitPath.Path, app.GitPath.ConfigFilename)
+		registeredAppPaths[id] = struct{}{}
+	}
+
+	if err := r.updateUnregisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update unregistered applications: %w", err)
+	}
+	if err := r.updateRegisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update registered applications: %w", err)
+	}
```

Comment:
Registered applications are more important than unregistered ones so let's change the order of updating.

Reply:
For sure 👍